src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -533,5 +538,154 @@ Json::Value Profile::ToJson() const
     JsonUtils::SetValueForKey(json, BellStyleKey, _BellStyle);
     JsonUtils::SetValueForKey(json, PixelShaderPathKey, _PixelShaderPath);
 
+    JsonUtils::SetValueForKey(json, EnvironmentKey, _EnvironmentVariables);
+
     return json;
 }
+
+namespace
+{
+    class VariablePair
+    {
+    public:
+        VariablePair(std::wstring rawValueIn) :
+            rawValue(std::move(rawValueIn))
+        {
+        }
+        VariablePair() = default;
+        virtual ~VariablePair() = default;
+        std::wstring rawValue;
+        std::wstring resolvedValue;
+    };
+
+    // Function Description:
+    // - Resolves a key's value from within envMap. Uses recursion to recursively resolve
+    //   referenced keys. envMap is updated as values are resolved so we don't double up on
+    //   work
+    // Arguments:
+    // - key: the key that may be in envMap or the environment (or unknown)
+    // - envMap: the map with all the keys, raw and resolved values
+    // - seenKeys: the list of keys that we have already seen when attempting to resolve the
+    //             current chain of keys. seenKeys needs to be passed in as a parameter so
+    //             that each layer in the recursion only has keys seen leading up to it
+    // Return Value:
+    // - The resolved value for key
+    std::wstring ResolveEnvironmentVariableValue(const std::wstring& key, std::map<std::wstring, VariablePair>& envMap, std::list<std::wstring> seenKeys = {})

```

Comment:
Everything in this method compares the names of environment variables case-sensitively. Is that by design?

Reply:
It was intended that way, however, I am struggling to find a use case where it needs case sensitivity. @zadjii-msft, should I make the environment variables case insensitive?

Even with a profile like git bash (which supports case sensitive env vars when running, they can't be applied through terminal:
```
          {
            "acrylicOpacity": 0.75,
            "background": "#000000",
            "closeOnExit": true,
            "colorScheme": "Campbell",
            "commandline": "\"%PROGRAMFILES%\\git\\bin\\bash.exe\" --login -i -l",
            "cursorColor": "#FFFFFF",
            "cursorShape": "bar",
            "fontFace": "Consolas",
            "fontSize": 10,
            "guid": "{00000000-0000-0000-0000-000000012345}",
            "historySize": 9001,
            "icon": "%PROGRAMFILES%\\git\\mingw64\\share\\git\\git-for-windows.ico",
            "name": "Git Bash",
            "padding": "0, 0, 0, 0",
            "snapOnInput": true,
            "useAcrylic": true,
            "environment": {
              "key": "a",
              "KEY": "b"
            }
          }
```

```
Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$ echo $KEY


Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$ echo $key
b

Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$

Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$ export MY_KEY=a

Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$ export my_key=b

Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$ echo $MY_KEY
a

Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$ echo $my_key
b

Chris@DESKTOP-R8BPF5L MINGW64 /c/WINDOWS/SysWOW64
$
```

Reply:
Yikes, this one I'm not sure about. I always assumed that env vars were case insensitive on Windows. So maybe the resolution here should be as well. That seems to make sense to me, but I don't think this is work blocking over tbh. @DHowett might have other thoughts though.

Reply:
Alas, yes.. I believe Windows' environment block expects to never find multiple case-differing copies of a single environment variable. We may want to keep Windows' tenets in the profile loader rather than Linux's.

Reply:
Are the case-insensitive string comparisons in GetEnvironmentVariable, ExpandEnvironmentStrings, etc. culture-dependent? I hope not.

Reply:
You can look in PT Run's environment helper class. I copied the entire behavior of windows. The helper behaves exactly the same as Windows does.

Are global profile environment variables planned?

Reply:
@htcfreek your insight might be appreciated over in #1125, where @miniksa is working on https://github.com/microsoft/terminal/blob/dev/miniksa/env/src/inc/til/env.h

Reply:
I can help with how windows behave. But I can't help eith coding. I don't know much about C++ language.